Nveto changes

WenzDaniel opened the pull request:
**What is the problem / what does the code in this PR do**
We add a pre-trigger time for the nVETORecorder, set the save_when of the Recorder to target, added a median based baseline estimator for large nveto pulses. See https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:wenz:comissioning:nveto:baseline:nveto_baseline for the latter.

jorana commented on straxen/plugins/nveto_recorder.py:
```diff
@@ -12,6 +12,8 @@
     strax.Option('coincidence_level_recorder_nv', type=int, default=4,
                  help="Required coincidence level."),
+    strax.Option('pre_trigger_time_nv', type=int, default=150,
+                 help="Pretrigger time before coincidence window."),
```
perhaps nice to add the unit, I assume ns but this is only a factor of 2 different from the number of samples.

jorana commented on straxen/plugins/nveto_recorder.py:
```diff
@@ -38,6 +40,7 @@ class nVETORecorder(strax.Plugin):
     parallel = 'process'
 
     rechunk_on_save = True
+    save_when = strax.SaveWhen.TARGET
```
👍

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -72,6 +83,46 @@ def compute(self, raw_records_coin_nv):
+        _, nextr = strax.record_links(records)
+
+        pulse_i = []
+        for i in np.where(records['record_i'] == 0)[0]:
```
Can we do this in a numba loop or vector operations? Perhaps performance is not affected but a while loop in a for loop of 2ns dt records looks ominous.

WenzDaniel replied:
See my comment in the wiki-note

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -72,6 +83,46 @@ def compute(self, raw_records_coin_nv):
+    
+    :param records: Records 
+    """
+    # Count number of pulses and create buffer:
```
Are you creating a buffer?

WenzDaniel replied:
Ah no this is old 👍

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -72,6 +83,46 @@ def compute(self, raw_records_coin_nv):
+    return records
+
+
+def _correct_baseline(records):
```
Can we numba-fy?

WenzDaniel replied:
See comment above

WenzDaniel replied:
Okay, numbafication brought a speed up of ~x1.5 - x2 so I added it (from 1.4 ms -> 800µs per saturation event). I also checked if the  `median_baseline` could become a bottleneck, but even if we would have 1k saturated events per run (which would indicate some problem). The processing only takes 700 ms.

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -21,6 +21,11 @@
              'Specify as a tuple of length n_nveto_pmts, or a number.'),
+    strax.Option(
+        'min_samples_alt_baseline_nv',
+        default=500, track=True,
```
I do see how this is going to create funny artifacts given the low levelness of this data. Also, a 1us long pulse is not really that long is it?

Perhaps you can also plot in https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:wenz:comissioning:nveto:baseline:nveto_baseline what the effect on short pulses would be?

WenzDaniel replied:
See my comments in the wiki.

jorana reviewed (changes requested):
Hi Daniel,
I have a couple of questions here:
https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:wenz:comissioning:nveto:baseline:nveto_baseline#questions

I also wonder about performance, it looks like some of these functions are slow python for loops. I'm mildly worried about some of these. 

Finally, could you version bump for bookkeeping.

WenzDaniel commented:
Okay since last time quite a few things changed. 
* We deactivated the alternative median baselining for the moment. Since there are some questions not resolved yet.
* We added a mini-analysis which allows to plot nveto pulses.
* We changed some of the default settings and added a pre-trigger time for the software trigger. 
* A new width computed via highest density regions was added. 

The PR now goes together with https://github.com/AxFoundation/strax/pull/384. All test past on my laptop.

jorana commented on straxen/analyses/nveto_plots.py:
```diff
@@ -0,0 +1,146 @@
+from matplotlib.backends.backend_pdf import PdfPages
+import os
+
+@straxen.mini_analysis(requires=('raw_records_nv',),
```
Hi Daniel, looks nice! Can you add an example of how a pulse should look? Perhaps it's also useful for the MV/TPC? Is it worth writing a wrapper function around this to allow it also to be used for raw_records_x?

WenzDaniel replied:
Ah yes, I forgot to link the example notebook sorry: https://github.com/XENONnT/analysiscode/blob/master/nveto/PulseDisplay.ipynb

> Perhaps it's also useful for the MV/TPC? Is it worth writing a wrapper function around this to allow it also to be used for raw_records_x?
That is actually a good point. We cannot use all features as for the nveto since e.g. the baseline estimated by the digitzer is not know in the other two cases. But let me see if I can make the required changes.

WenzDaniel replied:
Okay addressed your request and updated the MWE notebook.

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -21,6 +21,11 @@
              'Specify as a tuple of length n_nveto_pmts, or a number.'),
+    strax.Option(
+        'min_samples_alt_baseline_nv',
+        default=50000, track=True,
```
Is this activated or deactivated? It looks like it's on right, perhaps add a value (None, False) to the docstring to signify it's off?

WenzDaniel replied:
I changed it to a value which cannot be stratified, but you are right None or False would be more clean. Let me change it.

jorana commented on straxen/plugins/veto_pulse_processing.py:
```diff
@@ -72,6 +83,62 @@ def compute(self, raw_records_coin_nv):
+                inds += [ind]
+                ind = nextr[ind]
+                fail_counter += 1
+                assert fail_counter < 5000, 'Stuck in while-loop pulse is longer than 5000 fragments?!?'
```
yeah good idea to add something like this

jorana reviewed (approved):
Hi Daniel, thanks! Looks good!

Can you perhaps also add the plotting algos to the tests, very unlikely that something does not run but better safe than sorry :)

WenzDaniel commented:
Okay I addressed your comments. Tomorrow I will make the changes in https://github.com/AxFoundation/strax/pull/384

jorana commented on straxen/analyses/pulse_plots.py:
```diff
@@ -0,0 +1,178 @@
+def plot_pulses_mv(context, raw_records_mv, run_id, time_range,
+                    plot_hits=False, plot_median=False,
+                    max_plots=20, store_pdf=False, path=''):
+    plot_pulses(context, raw_records, run_id, time_range,
```
There is a copy paste error here in raw-records (_MV)

jorana reviewed (commented):
Hi Daniel, wonderful plots, very useful to have also for TPC!

WenzDaniel commented:
As soon we can merge https://github.com/AxFoundation/strax/pull/384. I think we can also merge this PR. I checked on my laptop and all tests pass.

jorana commented:
For some reason this is failing on my own laptop. Since the tests are not failing, it might be related to __pycache__ which I'm trying to figure out

WenzDaniel commented:
Okay let me check as well.

jorana reviewed (approved):
I approve but before you merge I think it would make sense to up the requirements of strax to accommodate the change in the hitlets dtype otherwise the dependency might be messy
